extension/src/lib/portBinding.js

```diff
@@ -13,51 +13,304 @@
+             - message id gets stored in a map with the promise to resolve/reject
+             -  receiver takes action of the message and returns a receipt ( includes the same message id )
+
+        receive - (receives a message)
+                - if it is a returned receipt
+                  - look into message map and resolve/reject corresponding promise
+                  - remove from map
+                - if it is a new message
+                  - take action on the message
+                  - send a return message (receipt) containing the same message id
+
+
+        Post Messages:
+        1) read/write message
+           - wait for receipt
+        2) read-receipt/write-receipt
+           - do not wait for receipt
+
+        Received Messages:
+        1) read/write message
+           - action, return receipt
+        2) read-receipt/write-receipt
+           - remove from message map
+
+        By default the following message types are handled.
+        - "UIO_PLUS_READ_REQUEST"
+        - "UIO_PLUS_READ_RECEIPT"
+        - "UIO_PLUS_WRITE_REQUEST"
+        - "UIO_PLUS_WRITE_RECEIPT"
+    */
+
     fluid.defaults("gpii.chrome.portBinding", {
         gradeNames: ["fluid.component"],
-        connectionName: "",
+        // Name of the port connection. Will be sent as the `name` when a port connection is created.
+        portName: "",
         members: {
-            lastIncomingMessage: {}
+            port: {
+                expander: {
+                    func: "{that}.setPort"
+                }
+            },
+            openRequests: {}
         },
         events: {
-            onIncomingMessage: null
+            onIncoming: null,
+            onIncomingRead: null,
+            onIncomingReadReceipt: null,
+            onIncomingWrite: null,
+            onIncomingWriteReceipt: null,
+            onDisconnect: null
+        },
+        // defines which event handles which a message type
+        // message types that aren't defined here are ignored.
+        messageForwardingMap: {
+            "UIO_PLUS_READ_REQUEST": "onIncomingRead",
+            "UIO_PLUS_READ_RECEIPT": "onIncomingReadReceipt",
+            "UIO_PLUS_WRITE_REQUEST": "onIncomingWrite",
+            "UIO_PLUS_WRITE_RECEIPT": "onIncomingWriteReceipt"
         },
         listeners: {
             "onCreate.bindPortEvents": "gpii.chrome.portBinding.bindPortEvents",
-            "onIncomingMessage.setLastIncomingMessage": {
-                listener: "fluid.set",
-                args: ["{that}", ["lastIncomingMessage"], "{arguments}.0"]
-            }
+            "onIncoming.filterMessages": {
+                listener: "gpii.chrome.portBinding.handleIncoming",
+                args: ["{that}", "{arguments}.0"]
+            },
+            "onIncomingRead.handle": {
+                listener: "{that}.handleMessage",
+                args: ["UIO_PLUS_READ_RECEIPT", "{that}.handleRead", "{arguments}.0"]
```

Comment:
As it stands an adapter of this component would have to override most of the implementation if they wanted to, e.g. share a port bus with another target. What I suggest is that
i) Within an expander you compute the inverse map of messageForwardingMap, and then
ii) Whenever you need to refer to an event name in the implementation, you refer to it via indirection on its corresponding event name - that is, these args would read "onIncomingRead"
and then in the implementation you would look up the corresponding port message type via `that.options.inverseMessageMap[eventName]`

Reply:
Well whichever message it is you should refer to it via a stable name, and the one you have to hand is the one defined in the messageForwardingMap